From 2fb49ebba0087ea3fbe697e0f52db3d714774c5c Mon Sep 17 00:00:00 2001
From: Sergei Miroshnichenko <s.miroshnichenko@yadro.com>
Date: Tue, 22 Sep 2020 12:45:19 +0300
Subject: [PATCH 38/57] powerpc/pci: Allow several pdns for the same PCI
 address

If a detached PCI device's phantom remains (bound by drivers), a new PCI
device may be found in the same slot during a PCI rescan.

Don't let the same pdn be linked to these both pci_devs, create a new one
instead.

Signed-off-by: Sergei Miroshnichenko <s.miroshnichenko@yadro.com>
---
 arch/powerpc/kernel/pci_dn.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/arch/powerpc/kernel/pci_dn.c b/arch/powerpc/kernel/pci_dn.c
index 111111111..111111111 100644
--- a/arch/powerpc/kernel/pci_dn.c
+++ b/arch/powerpc/kernel/pci_dn.c
@@ -82,6 +82,9 @@ struct pci_dn *pci_get_pdn_by_devfn(struct pci_bus *bus,
 		}
 	}
 
+	if (pci_has_flag(PCI_REASSIGN_ALL_BUS))
+		goto skip_slow_path;
+
 	if (!pdev_found)
 		pdev = NULL;
 
@@ -104,6 +107,7 @@ struct pci_dn *pci_get_pdn_by_devfn(struct pci_bus *bus,
 		}
 	}
 
+skip_slow_path:
 	return NULL;
 }
 
@@ -130,6 +134,9 @@ struct pci_dn *pci_get_pdn(struct pci_dev *pdev)
 	if (!parent)
 		return NULL;
 
+	if (pci_has_flag(PCI_REASSIGN_ALL_BUS))
+		goto skip_slow_path;
+
 	list_for_each_entry(pdn, &parent->child_list, list) {
 		if (pdn->busno == pdev->bus->number &&
 		    pdn->devfn == pdev->devfn) {
@@ -138,6 +145,7 @@ struct pci_dn *pci_get_pdn(struct pci_dev *pdev)
 		}
 	}
 
+skip_slow_path:
 	return pci_create_pdn_from_dev(pdev, parent);
 }
 

